File: Assets/Code/Outfit7/Raven/RavenOverseer.cc
#ifdef STARLITE
#include "RavenOverseer.h"

#include <Utility/RavenLog.h>

using namespace Starlite::Raven::Internal;

namespace Starlite {
    namespace Raven {
        const String RavenOverseer::Tag = "RavenOverseer";
        int RavenOverseer::s_MaximumSequenceWarmupFramesPerFrame = 1000;

        Queue<List<Ptr<RavenEventTriggerPoint>>*> RavenOverseer::s_TriggerPointListPool;
        Queue<List<Ptr<RavenPropertyComponent>>*> RavenOverseer::s_PropertyComponentListPool;
        Queue<Ptr<RavenEventTriggerPoint>> RavenOverseer::s_TriggerPointPool;
        Queue<Ptr<RavenSequence::IgnoredEventEntry>> RavenOverseer::s_IgnoredEventEntryPool;
        HashMap<String, UInt64> RavenOverseer::s_SequencePlayCountTable = HashMap<String, UInt64>(1024);
        int RavenOverseer::s_ActiveSequences = 0;
        int RavenOverseer::s_AllocatedFramesThisFrame = 0;
        UInt64 RavenOverseer::s_CurrentFrame = -1;
        int RavenOverseer::s_UsedIgnoredEventEntryCount = 0;
        int RavenOverseer::s_UsedPropertyComponentListCount = 0;
        int RavenOverseer::s_UsedTriggerPointCount = 0;
        int RavenOverseer::s_UsedTriggerPointListCount = 0;
        UInt64 RavenOverseer::s_MaximumSequencePlayCount = 0;

        RavenOverseer::constructor RavenOverseer::constructor_;

        void RavenOverseer::SetMaximumSequenceWarmupFramesPerFrame(int frames) {
            DebugAssert(frames > 0, "SetMaximumSequenceWarmupFramesPerFrame frames <= 0 (%d)", frames);
            s_MaximumSequenceWarmupFramesPerFrame = frames;
        }

        int RavenOverseer::GetAllocatedWarmupFrames(const RavenSequence* sequence, int maxFrames) {
            UInt64 frameCount = pTime->GetFrameCount();
            if (s_CurrentFrame != frameCount) {
                s_CurrentFrame = frameCount;
                s_AllocatedFramesThisFrame = 0;
            }

            if (s_AllocatedFramesThisFrame >= s_MaximumSequenceWarmupFramesPerFrame) {
                return 0;
            }

            UInt64 playCount = 0;
            auto it = s_SequencePlayCountTable.Find(sequence->GetSequenceName());
            if (it != s_SequencePlayCountTable.end()) {
                playCount = it->value;
            }

            double percentageOfMax = s_MaximumSequencePlayCount == 0 ? 1.0 : (playCount / (double)s_MaximumSequencePlayCount);
            double multiplier = 1 / (1.5 - percentageOfMax);
            int allocatedFrames = Math::Min((int)Math::Ceil(Math::Max(s_MaximumSequenceWarmupFramesPerFrame / s_ActiveSequences, 1) * multiplier), maxFrames);
            s_AllocatedFramesThisFrame += allocatedFrames;

#ifdef RAVEN_DEBUG
            pRavenLog->InfoT(Tag.GetCString(), "Allocated %d frames for %s. PlayCount: %llu MaximumPlayCount: %llu FramesLeft: %d", allocatedFrames, sequence->GetSequenceName(), playCount,
                             s_MaximumSequencePlayCount, MaximumSequenceWarmupFramesPerFrame - s_AllocatedFramesThisFrame);
#endif

            return allocatedFrames;
        }

        List<Ptr<RavenEventTriggerPoint>>* RavenOverseer::PopTriggerPointList() {
            const int addSize = 2000;

            if (s_TriggerPointListPool.Count() == 0) {
                pRavenLog->InfoT(Tag.GetCString(), "TrigerPointList pool empty. Adding %d elements. Total used %d.", addSize, s_UsedTriggerPointListCount);

                for (int i = 0; i < addSize; ++i) {
                    s_TriggerPointListPool.Enqueue(new List<Ptr<RavenEventTriggerPoint>>(8));
                }
            }

            List<Ptr<RavenEventTriggerPoint>>* list;
            Assert(s_TriggerPointListPool.Dequeue(&list), "Dequeue failed!");
            ++s_UsedTriggerPointListCount;

            return list;
        }

        List<Ptr<RavenPropertyComponent>>* RavenOverseer::PopPropertyComponentList() {
            const int addSize = 20;

            if (s_PropertyComponentListPool.Count() == 0) {
                pRavenLog->InfoT(Tag.GetCString(), "PropertyComponentList pool empty. Adding %d elements. Total used %d.", addSize, s_UsedPropertyComponentListCount);

                for (int i = 0; i < addSize; ++i) {
                    s_PropertyComponentListPool.Enqueue(new List<Ptr<RavenPropertyComponent>>(8));
                }
            }

            List<Ptr<RavenPropertyComponent>>* list;
            Assert(s_PropertyComponentListPool.Dequeue(&list), "Dequeue failed!");
            ++s_UsedPropertyComponentListCount;

            return list;
        }

        Ptr<RavenEventTriggerPoint> RavenOverseer::PopTriggerPoint(int frame, ERavenEventTriggerPointType type, Ptr<RavenEvent> sequencerEvent) {
            const int addSize = 2000;

            if (s_TriggerPointPool.Count() == 0) {
                pRavenLog->InfoT(Tag.GetCString(), "TrigerPoint pool empty. Adding %d elements. Total used %d.", addSize, s_UsedTriggerPointCount);

                for (int i = 0; i < addSize; ++i) {
                    s_TriggerPointPool.Enqueue(new RavenEventTriggerPoint());
                }
            }

            Ptr<RavenEventTriggerPoint> point;
            Assert(s_TriggerPointPool.Dequeue(&point), "Dequeue failed!");
            DebugAssert(point->IsPooled() == true, "TriggerPoint was not pooled when taking it from the pool!");

            point->Initialize(frame, type, sequencerEvent);
            point->SetIsPooled(false);
            ++s_UsedTriggerPointCount;

            return point;
        }

        Ptr<RavenSequence::IgnoredEventEntry> RavenOverseer::PopIgnoredEventEntry(Ptr<RavenEvent> evnt, int ignoreCount) {
            const int addSize = 16;

            if (s_IgnoredEventEntryPool.Count() == 0) {
                pRavenLog->InfoT(Tag.GetCString(), "IgnoredEventEntry pool empty. Adding %d elements. Total used %d.", addSize, s_UsedIgnoredEventEntryCount);

                for (int i = 0; i < addSize; ++i) {
                    s_IgnoredEventEntryPool.Enqueue(new RavenSequence::IgnoredEventEntry());
                }
            }

            Ptr<RavenSequence::IgnoredEventEntry> ignoredEventEntry;
            Assert(s_IgnoredEventEntryPool.Dequeue(&ignoredEventEntry), "Dequeue failed!");
            ignoredEventEntry->Initialize(evnt, ignoreCount);
            ++s_UsedIgnoredEventEntryCount;

            return ignoredEventEntry;
        }

        void RavenOverseer::PushIgnoredEventEntry(Ptr<RavenSequence::IgnoredEventEntry>& ignoredEventEntry) {
            if (ignoredEventEntry == nullptr) {
                return;
            }

            ignoredEventEntry->Reset();
            --s_UsedIgnoredEventEntryCount;

            s_IgnoredEventEntryPool.Enqueue(ignoredEventEntry);
        }

        void RavenOverseer::PushPropertyComponentList(List<Ptr<RavenPropertyComponent>>* list) {
            list->Clear();
            --s_UsedPropertyComponentListCount;

            s_PropertyComponentListPool.Enqueue(list);
        }

        void RavenOverseer::PushTriggerPoint(Ptr<RavenEventTriggerPoint>& point) {
            if (point == nullptr) {
                return;
            }

            DebugAssert(point->IsPooled() == false, "TriggerPoint was already pooled when pushing it back to the pool!");

            point->Reset();
            point->SetIsPooled(true);
            --s_UsedTriggerPointCount;

            s_TriggerPointPool.Enqueue(point);
        }

        void RavenOverseer::PushTriggerPointList(List<Ptr<RavenEventTriggerPoint>>* list) {
            if (list == nullptr) {
                return;
            }

            for (auto& el : *list) {
                PushTriggerPoint(el);
            }

            list->Clear();
            --s_UsedTriggerPointListCount;

            s_TriggerPointListPool.Enqueue(list);
        }

        void RavenOverseer::RegisterSequence(Ptr<RavenSequence> sequence) {
            ++s_ActiveSequences;
        }

        void RavenOverseer::RegisterSequencePlay(Ptr<RavenSequence> sequence, bool forcefulWarmup) {
            String sequenceName = sequence->GetSequenceName();
            UInt64 addPlayCount = 1;

            if (forcefulWarmup) {
                addPlayCount = 10;
            }

            UInt64 playCount = 0;
            auto it = s_SequencePlayCountTable.Find(sequenceName);
            if (it != s_SequencePlayCountTable.end()) {
                playCount += addPlayCount;
            } else {
                playCount = addPlayCount;
            }

            s_SequencePlayCountTable[sequenceName] = playCount;

            if (playCount > s_MaximumSequencePlayCount) {
                s_MaximumSequencePlayCount = playCount;
            }
        }

        void RavenOverseer::UnregisterSequence(Ptr<RavenSequence> sequence) {
            --s_ActiveSequences;
        }

        void RavenOverseer::InitRavenEventTriggerPointListPool() {
            const int poolSize = 20000;

            s_TriggerPointListPool = Queue<List<Ptr<RavenEventTriggerPoint>>*>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_TriggerPointListPool.Enqueue(new List<Ptr<RavenEventTriggerPoint>>(8));
            }
        }

        void RavenOverseer::InitRavenEventTriggerPointPool() {
            const int poolSize = 20000;

            s_TriggerPointPool = Queue<Ptr<RavenEventTriggerPoint>>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_TriggerPointPool.Enqueue(new RavenEventTriggerPoint());
            }
        }

        void RavenOverseer::InitRavenIgnoredEventEntryPool() {
            const int poolSize = 20;

            s_IgnoredEventEntryPool = Queue<Ptr<RavenSequence::IgnoredEventEntry>>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_IgnoredEventEntryPool.Enqueue(new RavenSequence::IgnoredEventEntry());
            }
        }

        void RavenOverseer::InitRavenPropertyComponentListPool() {
            const int poolSize = 100;

            s_PropertyComponentListPool = Queue<List<Ptr<RavenPropertyComponent>>*>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_PropertyComponentListPool.Enqueue(new List<Ptr<RavenPropertyComponent>>(8));
            }
        }
    } // namespace Raven
} // namespace Starlite

#endif

File: Assets/Code/Outfit7/Raven/RavenOverseer.cs
using Starlite.Raven.Internal;
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Starlite.Raven {

    public static class RavenOverseer {
        private const string Tag = "RavenOverseer";

        private static int s_MaximumSequenceWarmupFramesPerFrame = 1000;

        private static Queue<RavenEventTriggerPoint> s_TriggerPointPool;
        private static int s_UsedTriggerPointCount = 0;

        private static Queue<List<RavenEventTriggerPoint>> s_TriggerPointListPool;
        private static int s_UsedTriggerPointListCount = 0;

        private static Queue<RavenSequence.IgnoredEventEntry> s_IgnoredEventEntryPool;
        private static int s_UsedIgnoredEventEntryCount;

        private static Queue<List<RavenPropertyComponent>> s_PropertyComponentListPool;
        private static int s_UsedPropertyComponentListCount = 0;

        private static readonly Dictionary<string, ulong> s_SequencePlayCountTable;
        private static ulong s_MaximumSequencePlayCount;
        private static int s_ActiveSequences;
        private static int s_AllocatedFramesThisFrame;
        private static int s_CurrentFrame;

        static RavenOverseer() {
            s_CurrentFrame = -1;
            s_SequencePlayCountTable = new Dictionary<string, ulong>(1024);
            InitRavenEventTriggerPointPool();
            InitRavenEventTriggerPointListPool();
            InitRavenIgnoredEventEntryPool();
            InitRavenPropertyComponentListPool();
        }

        #region Utility

        public static void SetMaximumSequenceWarmupFramesPerFrame(int frames) {
            if (frames <= 0) {
                return;
            }
            s_MaximumSequenceWarmupFramesPerFrame = frames;
        }

        #endregion Utility

        #region SequenceWarmup

        public static void RegisterSequence(RavenSequence sequence) {
#if UNITY_EDITOR
            if (!Application.isPlaying) {
                return;
            }
#endif
            ++s_ActiveSequences;
        }

        public static void UnregisterSequence(RavenSequence sequence) {
#if UNITY_EDITOR
            if (!Application.isPlaying) {
                return;
            }
#endif
            --s_ActiveSequences;
        }

        public static void RegisterSequencePlay(RavenSequence sequence, bool forcefulWarmup) {
            var sequenceName = sequence.Name;
            ulong addPlayCount = 1;

            if (forcefulWarmup) {
                addPlayCount = 10;
            }

            ulong playCount = 0;
            if (!s_SequencePlayCountTable.TryGetValue(sequenceName, out playCount)) {
                playCount = addPlayCount;
            } else {
                playCount += addPlayCount;
            }

            s_SequencePlayCountTable[sequenceName] = playCount;

            if (playCount > s_MaximumSequencePlayCount) {
                s_MaximumSequencePlayCount = playCount;
            }
        }

        public static int GetAllocatedWarmupFrames(RavenSequence sequence, int maxFrames) {
#if UNITY_EDITOR
            if (!Application.isPlaying) {
                return maxFrames;
            }
#endif

            if (s_CurrentFrame != Time.frameCount) {
                s_CurrentFrame = Time.frameCount;
                s_AllocatedFramesThisFrame = 0;
            }

            if (s_AllocatedFramesThisFrame >= s_MaximumSequenceWarmupFramesPerFrame) {
                return 0;
            }

            ulong playCount;
            if (!s_SequencePlayCountTable.TryGetValue(sequence.Name, out playCount)) {
                playCount = 0;
            }

            var percentageOfMax = s_MaximumSequencePlayCount == 0 ? 1.0 : (playCount / (double)s_MaximumSequencePlayCount);
            var multiplier = 1 / (1.5 - percentageOfMax);
            var allocatedFrames = Math.Min((int)Math.Ceiling(Math.Max(s_MaximumSequenceWarmupFramesPerFrame / s_ActiveSequences, 1) * multiplier), maxFrames);
            s_AllocatedFramesThisFrame += allocatedFrames;
#if RAVEN_DEBUG
            RavenLog.DebugT(Tag, "Allocated {0} frames for {1}. PlayCount: {2} MaximumPlayCount: {3} FramesLeft: {4}", allocatedFrames, sequence.Name, playCount, s_MaximumSequencePlayCount, MaximumSequenceWarmupFramesPerFrame - s_AllocatedFramesThisFrame);
#endif

            return allocatedFrames;
        }

        #endregion SequenceWarmup

        #region Pools

        public static RavenEventTriggerPoint PopTriggerPoint(int frame, ERavenEventTriggerPointType type, RavenEvent sequencerEvent) {
            const int addSize = 2000;

            if (s_TriggerPointPool.Count == 0) {
                RavenLog.DebugT(Tag, "TrigerPoint pool empty. Adding {0} elements. Total used {1}.", addSize, s_UsedTriggerPointCount);

                for (int i = 0; i < addSize; ++i) {
                    s_TriggerPointPool.Enqueue(new RavenEventTriggerPoint());
                }
            }

            var point = s_TriggerPointPool.Dequeue();
#if RAVEN_DEBUG
            RavenAssert.IsTrue(point.IsPooled == true, "TriggerPoint was not pooled when taking it from the pool!");
#endif
            point.Initialize(frame, type, sequencerEvent);
            point.IsPooled = false;
            ++s_UsedTriggerPointCount;

            return point;
        }

        public static void PushTriggerPoint(RavenEventTriggerPoint point) {
            if (point == null) {
                return;
            }

#if RAVEN_DEBUG
            RavenAssert.IsTrue(point.IsPooled == false, "TriggerPoint was already pooled when pushing it back to the pool!");
#endif
            point.Reset();
            point.IsPooled = true;
            --s_UsedTriggerPointCount;

            s_TriggerPointPool.Enqueue(point);
        }

        private static void InitRavenEventTriggerPointPool() {
            const int poolSize = 20000;

            s_TriggerPointPool = new Queue<RavenEventTriggerPoint>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_TriggerPointPool.Enqueue(new RavenEventTriggerPoint());
            }
        }

        public static List<RavenEventTriggerPoint> PopTriggerPointList() {
            const int addSize = 2000;

            if (s_TriggerPointListPool.Count == 0) {
                RavenLog.DebugT(Tag, "TrigerPointList pool empty. Adding {0} elements. Total used {1}.", addSize, s_UsedTriggerPointListCount);

                for (int i = 0; i < addSize; ++i) {
                    s_TriggerPointListPool.Enqueue(new List<RavenEventTriggerPoint>(8));
                }
            }

            var list = s_TriggerPointListPool.Dequeue();
            ++s_UsedTriggerPointListCount;

            return list;
        }

        public static void PushTriggerPointList(List<RavenEventTriggerPoint> list) {
            if (list == null) {
                return;
            }

            for (int i = 0; i < list.Count; ++i) {
                PushTriggerPoint(list[i]);
            }

            list.Clear();
            --s_UsedTriggerPointListCount;

            s_TriggerPointListPool.Enqueue(list);
        }

        private static void InitRavenEventTriggerPointListPool() {
            const int poolSize = 20000;

            s_TriggerPointListPool = new Queue<List<RavenEventTriggerPoint>>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_TriggerPointListPool.Enqueue(new List<RavenEventTriggerPoint>(8));
            }
        }

        public static RavenSequence.IgnoredEventEntry PopIgnoredEventEntry(RavenEvent evnt, int ignoreCount) {
            const int addSize = 16;

            if (s_IgnoredEventEntryPool.Count == 0) {
                RavenLog.DebugT(Tag, "IgnoredEventEntry pool empty. Adding {0} elements. Total used {1}.", addSize, s_UsedIgnoredEventEntryCount);

                for (int i = 0; i < addSize; ++i) {
                    s_IgnoredEventEntryPool.Enqueue(new RavenSequence.IgnoredEventEntry());
                }
            }

            var ignoredEventEntry = s_IgnoredEventEntryPool.Dequeue();
            ignoredEventEntry.Initialize(evnt, ignoreCount);
            ++s_UsedIgnoredEventEntryCount;

            return ignoredEventEntry;
        }

        public static void PushIgnoredEventEntry(RavenSequence.IgnoredEventEntry ignoredEventEntry) {
            if (ignoredEventEntry == null) {
                return;
            }

            ignoredEventEntry.Reset();
            --s_UsedIgnoredEventEntryCount;

            s_IgnoredEventEntryPool.Enqueue(ignoredEventEntry);
        }

        private static void InitRavenIgnoredEventEntryPool() {
            const int poolSize = 20;

            s_IgnoredEventEntryPool = new Queue<RavenSequence.IgnoredEventEntry>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_IgnoredEventEntryPool.Enqueue(new RavenSequence.IgnoredEventEntry());
            }
        }

        public static List<RavenPropertyComponent> PopPropertyComponentList() {
            const int addSize = 20;

            if (s_PropertyComponentListPool.Count == 0) {
                RavenLog.DebugT(Tag, "PropertyComponentList pool empty. Adding {0} elements. Total used {1}.", addSize, s_UsedPropertyComponentListCount);

                for (int i = 0; i < addSize; ++i) {
                    s_PropertyComponentListPool.Enqueue(new List<RavenPropertyComponent>(8));
                }
            }

            var list = s_PropertyComponentListPool.Dequeue();
            ++s_UsedPropertyComponentListCount;

            return list;
        }

        public static void PushPropertyComponentList(List<RavenPropertyComponent> list) {
            if (list == null) {
                return;
            }

            list.Clear();
            --s_UsedPropertyComponentListCount;

            s_PropertyComponentListPool.Enqueue(list);
        }

        private static void InitRavenPropertyComponentListPool() {
            const int poolSize = 100;

            s_PropertyComponentListPool = new Queue<List<RavenPropertyComponent>>(poolSize);
            for (int i = 0; i < poolSize; ++i) {
                s_PropertyComponentListPool.Enqueue(new List<RavenPropertyComponent>(8));
            }
        }

        #endregion Pools
    }
}